### ISSUE TITLE
Pyright does not respect `dataclass_transform` argument `field_specifiers` when using different types.

### ISSUE THREAD

**Issue by imgurbot12** [microsoft/pyright#5102]
**Describe the bug**

When using the new python decorator `dataclass_transform` pyright still returns errors when specifying a field of a type already declared in the `field_specifiers` argument.

**To Reproduce**

Use a custom type/function as defined in `field_specifiers` of the `dataclass_transform` decorator. _(see example code below)_

**Expected behavior**
Pyright should respect types/functions declared within the field-specifiers which declares what types are allowed for fields.

**Screenshots or Code**

Example Code:
```python
from typing import List, dataclass_transform

class Field:
    pass

def field(*args, **kwargs) -> Field:
    return Field()

@dataclass_transform(field_specifiers=(Field, field))
def test(cls):
    return cls

@test
class Foo:
    a: int
    b: List[str] = field(default_factory=list)
```
Example Output:
```bash
$ pyright test.py 
<filepath>/test.py
  <filepath>/test.py:16:20 - error: Expression of type "Field" cannot be assigned to declared type "List[str]"
    "Field" is incompatible with "List[str]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
This error is raised using the pyright command-line: `pyright test.py`

**Comment by erictraut** [comment]
Type checkers assume that the functional form of a field specifier will return the type of the field or `Any`, not a `Field` object. This is the convention that was established with stdlib's `dataclass.field` function. Refer to `dataclasses.pyi` for an example.
